### ISSUE TITLE
tfdbg strange things about detecting NAN weights in network

### ISSUE THREAD

**Issue by Lan1991Xu** [tensorflow/tensorflow#8410]
### What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?
Hi all, I just meet the NAN weight in the training for my network, I have check the loss and see whether there is like the form of 0/0, or log(0) and changing the learning rate but it still meet the problem when I try in the 200 steps. So i use tfdbg to debug, it is really nice tool. But when I use it, it running only several steps and stop. And I print this node out, this node has so many numbers. when I use pt node, and there is so many numbers to detect, I am not find the inf and nan. Besides, the weights i even not change it, so what cause this problem
### Environment info
Operating System:
Ubuntu 14:04
Cuda 8.0
cuDNN5.1
Installed version of CUDA and cuDNN: 

this is output in the screen--- run-end: run #103: fetch: t_DQN/dqn_q/BiasAdd:0; 2 feeds ----------------------------------------------------
Tensor "encoder/conv0/Conv2D:0:DebugIdentity":
  dtype: float32
  shape: (8, 149, 149, 32)

array([[[[ -2.31261123e-02,  -2.28441790e-01,  -1.15194634e-01,   1.87772080e-01,   1.52234808e-01,
            9.16329771e-02,  -1.40365764e-01,   8.70720018e-04,   1.98106512e-01,  -3.14010084e-02,
            5.08706868e-02,  -1.47755712e-01,   1.76122814e-01,   4.00081761e-02,  -1.88322872e-01,
           -1.56596974e-01,  -1.78046972e-01,  -1.91862926e-01,  -1.90227181e-01,  -6.63982928e-02,
            1.77524462e-01,  -7.36486465e-02,  -6.05005510e-02,   1.09997056e-01,   9.24435183e-02,
            4.22568992e-05,  -1.06559973e-03,   9.99473780e-02,  -1.73573643e-02,   1.16514668e-01,
           -1.51743174e-01,   2.23651499e-01],
         [ -2.14941762e-02,  -2.30754048e-01,  -1.18553132e-01,   1.89629942e-01,   1.53854966e-01,

**Comment by Lan1991Xu** [comment]
@caisq , thanks, if you have time to check it.

**Comment by caisq** [comment]
@Lan1991Xu,

tfdbg CLI has two features that might be useful to you here.
1) Regular expression search, similar to the Linux `less` UI. For example, if you want to search for `inf` and `nan`, you can scroll to the top with the `Home` key and then do:
```
tfdbg> /(inf|nan)
```
The CLI will try to find `inf` or `nan` and if any is found, scroll to the first instance. If any match is found, you can continue to the next ones by using `tfdbg> /`, just like in `less`.

2) You can slice large tensors by using numpy style slicing, e.g.,
```
tfdbg> pt encoder/conv0/Conv2D:0:DebugIdentity[0, :, :, :]
```

If you have more questions, let me know.

**Comment by caisq** [comment]
There is actually another potentially useful feature: the `-r` option of the `pt` command. It allows you to count how many elements a Tensor falls into a certain range. The range can be a real number or `inf`. If you do:
```
tfdbg> pt encoder/conv0/Conv2D:0:DebugIdentity -r [-inf,inf]
```
the CLI will count how many elements are >=-inf and <inf. So if you there is one NaN, the element will not be counted. You'll see in the first line on the top something like:
999 of 1000 elements (99.90%)

**Comment by Lan1991Xu** [comment]
Hi Cai
Thanks for your very useful suggestion, it seems in this pt encoder/conv0/Conv2D:0 -r [-inf,inf] really have NAN, the output of the number is small than predefined number.  However, the strange things is that I use this to extract my features and test, the performance is really good. Actually, the weight NAN really influence my performance is another network(DQN network). I use the first network to extract feature(the weight is fixed). Can i directly check the NAN in my DQN network, ignoring the NAN number in the feature extraction?

**Comment by Lan1991Xu** [comment]
By the way, you means using the command lt. I guess if I directly use run -f has_inf_or_nan, the output of dumped tensors are all related inf and nan problems, so I have no need to do that?

**Comment by caisq** [comment]
@Lan1991Xu 

Regarding your first question, if the nodes in your DQN network have a common naming pattern, maybe you can use the `-n` option of `lt`, such as:
```
tfdbg> lt -f has_inf_or_nan -n "^DQN/.*"
```

This will ignore all the Tensors with NaNs and Infs whose name do not begin with DQN.

If you want to write a tensor_filter (i.e., conditional breakpoint between runs) similar to tf_debug.has_inf_or_nan, but checks only the "^DQN/.*" notes, you can do:
```
def has_inf_or_nan_in_dqn(datum, tensor):
  return datum.node_name.startswith("DQN") and tf_debug.has_inf_or_nan(tensor)

sess = tf_debug.LocalCLIDebugWrapperSession(sess)
sess.add_tensor_filter("has_inf_or_nan_in_dqn", has_inf_or_nan_in_dqn)
```

Then when you do `run -f has_inf_or_nan_in_dqn`, tfdbg will only break if the DQN network part of your model contains NaNs or Infs.

Regarding your second question, yes, if you use the `run -f ...` command, the initial screen automatically does `lt -f ...` for you.

**Comment by Lan1991Xu** [comment]
Thanks caisq!I just find the bug, it takes me about two days. tfdbg is really powerful tools, I find it is indeed cause by my feature extraction when i move action on images!  This tools really amazing for us to debug tensorflow. By the way, when I use this tfdbg, it seems when I need to save the model and also using tfdbg, it will report: ufunc"is nan" not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule"safe". I guess it is  the filter problem, not about the tfdbg itself. Any suggestions about define this filter to avoid this problems?

**Comment by caisq** [comment]
@Lan1991Xu Glad it worked for you.

Regarding the ufunc error. Can you copy and paste the entire error message and stack trace if it is not too inconvenient for you? Thanks.

**Comment by Lan1991Xu** [comment]
Hi caisq
The report error is :   ufunc"is nan" not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule"safe"
When I use filter "has inf or nan", I guess it is some conflict about input file or save the data. So when I temporary remove  the save data "  self.saver.save(self.sess, self.model_dir + 'snapshot', global_step = step) " and load weights before I add "tf_debug.LocalCLIDebugWrapperSession(self.sess)", it works.

**Comment by Lan1991Xu** [comment]
However, for some code, like test on Inception V3 for slim, I can not simply change the load weights or data. It is difficult for me to debug on inception V3 model on NAN. The problem I guess is related to the specific filter.  After use this tool, it is really convenient for me to debug tensorflow, and I also find how to place "self.sess= tf_debug.LocalCLIDebugWrapperSession(self.sess)
		#self.sess.add_tensor_filter("has_inf_or_nan", tf_debug.has_inf_or_nan)" is also important.
Any suggestions about this. Cause i find I also meet the problem " Error in tensorflow debugger (tfdbg) while executing session run call in child thread", but only change the place of this line, it works.

**Comment by Lan1991Xu** [comment]
Solve it. It is nice to use this.

**Comment by caisq** [comment]
Thanks for checking, @skye. Thanks for confirming, @Lan1991Xu .
